### PR TITLE
DateTimeField implementation

### LINKED ISSUE #1438: In need of the Datetime field
I would like to see the Datetime field as option when creating fields.
Probably somewhere on the roadmap, but just to be sure...

I'm creating blogposts and I'm not able to sort on a (publish) date?

### PR DESCRIPTION
**Pull request by agriffard**
Fixes #1438

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Fields/DateTimeField.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+using System;
+using OrchardCore.ContentManagement;
+
+namespace OrchardCore.ContentFields.Fields
+{
+    public class DateTimeField : ContentField
+    {
+        public DateTime? Value { get; set; }
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Fields/DateTimeField.cs]: I understand that a field could not be required and have a null value. But I don't understand how the current code would support that. Does it work if nothing is set in the editor, does the updater correctly assigns null?

Also related, let's say I set a value and save, and the value is not required, can I erase the value somehow?

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Fields/DateTimeField.cs]: I was able to reset the value to nothing using the local editor.

**Comment by sebastienros** [comment]
It still needs to handle time zone conversion.
I tried to change the time zone of the site and it didn't affect the value that is edited, it should.
I also tried to access it in liquid and it worked fine: `{{ Model.ContentItem.Content.Appointment.Scheduled.Value | local }}`

**Comment by agriffard** [comment]
I merged the dev branch and now the `| local` works.

@Skrypt : Please help to understand the best way to store the DateTime value as UTC and use the current site timezone (and the user timezone) to display the date accordingly.
Do I need to wait your Timezone PR to be merged first?

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/DateTimeFieldDriver.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            .Location("Content")
+            .Location("SummaryAdmin", "");
+            ;
+        }
+
+        public override IDisplayResult Edit(DateTimeField field, BuildFieldEditorContext context)
+        {
+            return Initialize<EditDateTimeFieldViewModel>("DateTimeField_Edit", model =>
+            {
+                model.Value = field.Value;
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/DateTimeFieldDriver.cs]: Convert from UTC to local time zone.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/DateTimeFieldDriver.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                model.Field = field;
+                model.Part = context.ContentPart;
+                model.PartFieldDefinition = context.PartFieldDefinition;
+            });
+        }
+
+        public override async Task<IDisplayResult> UpdateAsync(DateTimeField field, IUpdateModel updater, UpdateFieldEditorContext context)
+        {
+            await updater.TryUpdateModelAsync(field, Prefix, f => f.Value);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/DateTimeFieldDriver.cs]: Create a new `EditDateTimeFieldViewModel` instance and bind the `Value` property.
Then convert it to UTC (from the current time zone) and assign to `field.Value`.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/DateTimeFieldDriver.cs (hunk 2 above)
Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/DateTimeFieldDriver.cs]: `model.Value = (await _localClock.ConvertToLocal(field.Value)).DateTime;`

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/DateTimeFieldDriver.cs (hunk 3 above)
Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Drivers/DateTimeFieldDriver.cs]: `field.Value = new DateTimeOffset(model.Value,  _localClock.GetTimeZoneAsync()).ToUniversal().DateTime;`

**Comment by sebastienros** [comment]
All fixed and tested. Please check it works as expected an merge.

From a template we can use `{{ Model.LocalDateTime }}` or `{{ Model.Field.Value | local }}` which will return a local datetime. `Model.Field.Value` will return the UTC version.

**Pull request merged**